### PR TITLE
Improve quantity check for Product and Cart

### LINKED ISSUE #267: Handle cart errors better
If something goes wrong during adding product to cart, empty modal is displayed.

We should get inspired by @Oksydan's theme and handle these situations better. ;-)

https://github.com/Oksydan/is_shoppingcart

### PR DESCRIPTION
**Pull request by mparvazi**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Fixes response on "Add to cart" when wanted quantity is more than available quantity and improves quantity check in Cart Controller.
| Type?             | improvement
| Category?         | FO 
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/hummingbird/issues/267
| Related PRs       | no
| How to test?      | Stock management must be enabled for products. Need to have an attribute with specific quantity for a Product. In Product page and Cart page try to set quantity more than available quantity. if the desired qty is > the available qty, we return an error a little more detailed and on the core.js side we emit the error to display a nice message in the themes
| Possible impacts? | In Product page add-to-cart action. In Cart page update quantity for product.

# Before this PR

### Cart page in Classic

https://user-images.githubusercontent.com/85633460/178004467-0fe7098b-9668-4d09-b506-5271bc11db45.mp4

### Cart page in Hummingbird

https://user-images.githubusercontent.com/85633460/178004528-f2405264-3e63-4334-a4bf-2b3672e2b107.mp4

### Product page in Classic

https://user-images.githubusercontent.com/85633460/178004617-021e9c13-5486-4c50-a794-52b938909d0a.mp4

### Product page in Hummingbird

https://user-images.githubusercontent.com/85633460/178004715-9c11a719-da35-4f17-bcb5-7c3cccb5bf64.mp4

# After this PR
### Cart page in Classic

https://user-images.githubusercontent.com/85633460/178004861-ba1fa4eb-9403-4d61-b133-5e8525c78660.mp4

### Cart page in Hummingbird

https://user-images.githubusercontent.com/85633460/178004880-abfb6fdb-eed0-41e8-a588-12d8de4ec3ef.mp4

### Product page in Classic

Nothing changed!

### Product page in Hummingbird
Now we have an option to show error:
 
167398589-065fb64d-5e61-4d9b-9923-da33e48e6126

https://user-images.githubusercontent.com/85633460/178004994-8d02612c-2fec-49f6-8552-271df968771e.mp4

### CODE REVIEW

**Comment by mparvazi** [comment]
@NeOMakinG @Hlavtox @SharakPL 

Any comments are welcome!
Btw let's to get rid of this weird response:

167361811-c6408174-f7bf-4112-9ed6-eb2dfba591b1

**Review by Hlavtox — DISMISSED** [review]
You are the man!!

**Comment by SharakPL** [comment]
Yeah, all good. Although I would change this alert:
https://github.com/PrestaShop/PrestaShop/blob/05ddf0bcb56bd74d3f69fe3126b799a5f58a8053/controllers/front/CartController.php#L435
because it's rather misleading. Like it's just been sold out and you can't buy it anymore. It should be specific about the quantity, not product availability:

> Requested number exceeds the available quantity of this product

and since this alert remains even after the number is automatically lowered then it should be specific about available quantity:

> Requested number exceeds the available quantity **(5)** of this product

TBH I don't care about Classic anymore, but it can generate problems there, because the alert pops out before the cart, not in product line like in hummingbird, so when there's a lot of products in the cart causing a scroll and someone exceeds the available amount on the last item in the cart, then this alert won't be even visible at first.
It would be better to get rid of this allert completely and simply show available items:
obraz
Then just paint it red if someone goes higher and maybe add a popover to the input with alert text.

**Comment by mparvazi** [comment]
@SharakPL 

I think we should not display remain quantity on response to the user (e.g. Available: 5 ),
maybe when we are showing this quantity (that confirms customer can order this quantity),
someone else has started buying this quantity and order has completed,
so this quantity is not valid anymore.

It's better to check this quantity on-demand and let's back-end controls this process.

In Classic theme, inside Product page, this value will be checked every time customer changes the value,
but there is 2 problems about that:
1. Performance (too request to the back-end)
2. Fast clicking bug

Btw if it's important to show available quantity, we need to find a proper message, to show that there is no guarantee customer can order by this quantity.

**Comment by SharakPL** [comment]
> I think we should not display remain quantity on response to the user (e.g. Available: 5 ), maybe when we are showing this quantity (that confirms customer can order this quantity), someone else has started buying this quantity and order has completed, so this quantity is not valid anymore.
> 
Exactly the same can happen here, right? And if available quantity would be checked on each input change then there's no problem to also update "Available: X" value. I'm thinking about something like this:

quantity

No alerts somewhere on the page. All feedbacks right there with quantity form.

Of course this is just an option for the future. For now this fix will do.

**Comment by mparvazi** [comment]
> > I think we should not display remain quantity on response to the user (e.g. Available: 5 ), maybe when we are showing this quantity (that confirms customer can order this quantity), someone else has started buying this quantity and order has completed, so this quantity is not valid anymore.
> 
> Exactly the same can happen here, right? And if available quantity would be checked on each input change then there's no problem to also update "Available: X" value. I'm thinking about something like this:
> 
> quantity [ ](https://user-images.githubusercontent.com/5007456/178152627-fb134b4d-3762-4c5b-b9c7-a433414bce40.gif) quantity [ ](https://user-images.githubusercontent.com/5007456/178152627-fb134b4d-3762-4c5b-b9c7-a433414bce40.gif)
> 
> No alerts somewhere on the page. All feedbacks right there with quantity form.

I think this process currently is used on Classic theme.
This PR doesn't block this ability (I'll try to check again).

**Comment by mparvazi** [comment]
I tested again, Classic theme uses `product.js` in core to check quantity wanted availability (product refresh request).
So there is no conflict with this PR.

I think we just need to send  'available quantity' from the back-end with response error:

Screenshot from 2022-07-10 22-38-54


https://user-images.githubusercontent.com/85633460/178157339-027cfc50-a355-467d-aead-3685986df5b3.mp4

**Comment by Hlavtox** [comment]
It's sad that the community needs to fix add to cart functionality of an ecommerce software 🤣

**Comment by SharakPL** [comment]
@mparvazi that's great! One small issue though:
obraz

There's still 300 available, not 100, because the order hasn't been placed yet. It should match the maximum number for the input, so it should say:

> The available quantity of this product is 300

or even clearer:

> The available quantity of this product is 300. You can add up to 100 more.

or maybe this will be better:

> Maximum quantity of this product available for purchase is currently 300.

**Comment by mparvazi** [comment]
@SharakPL 

I think this one is better:
>  Maximum quantity of this product available for purchase is currently 300.

On the other hand this one is more useful on Product page:
> The available quantity of this product is 300. You can add up to 100 more.

Btw the error message must be same for Cart and Product page.

**Comment by SharakPL** [comment]
@NeOMakinG @kpodemski wdyt?

**Comment by mparvazi** [comment]
https://user-images.githubusercontent.com/85633460/178220473-8d99c6e4-c52a-4c79-a07a-5daf5d45e644.mp4

**Comment by mparvazi** [comment]
https://user-images.githubusercontent.com/85633460/178220473-8d99c6e4-c52a-4c79-a07a-5daf5d45e644.mp4

**Comment by SharakPL** [comment]
> On the other hand this one is more useful on Product page:
> 
> > The available quantity of this product is 300. You can add up to 100 more.

I agree this would be better. Otherwise product page should contain information somewhere near add to cart button on quantity of this product that's already in the cart.

Also I'm not sure how to translate _"available purchase order quantity"_. What do you mean by "purchase order quantity"? IMO it should clear and obvious:
- The available quantity of this product is X. You can add up to Y more to your cart.
- The quantity of this product available for purchase is X. You can add up to Y more to your cart.

**Comment by NeOMakinG** [comment]
> It's sad that the community needs to fix add to cart functionality of an ecommerce software 🤣

This is what open source is made for! It's not like the core team was doing nothing for the software at all :/

**Review by NeOMakinG — APPROVED** [review]
This is simple and awesome 🚀

**Comment by SharakPL** [comment]
@NeOMakinG what about the alert text? It should be more clear imo

**Comment by NeOMakinG** [comment]
@PrestaShop/product-team @Julievrz We need your input there, if you've some time, it would be awesome 🙏

**Comment by Julievrz** [comment]
What about something like: "There are only X products left in stock." 

The only question I have is: Is the number of available products the same as the number of products the user can add to his cart? If yes, we can go with this wording. 

Otherwise, we should insist on the second point (because what's the point of knowing there are 50 products left if I can only add 3?), with something like: "You can only add 3 products to your cart."

Let me know what you think. :)

**Comment by SharakPL** [comment]
That's why I would rather be specific about it with something like _"The quantity of this product available for purchase is X. You can add up to Y more to your cart."_ This way user has clear information on max items that can be added to the cart depending on number of items already in the cart.

If there's no items of the product (combination) in the cart yet then it could shortened to _"The quantity of this product available for purchase is X"_. Would probably be easier that way to make 2 separate sentences to translate.

**Comment by mparvazi** [comment]
@Julievrz @SharakPL 

It's possible to have different wording for Product page and Shopping cart.

My suggestion:

Product page (when this combination already exists in cart):
`There is/are only X product(s) left in stock. You can add Y product(s) to your cart.`

Shopping cart & Product page (when this combination already NOT exists in cart):
`There is/are only X product(s) left in stock.`

**Comment by SharakPL** [comment]
> Product page (when this combination already exists in cart): `There are only X product(s) left in stock. You can add Y product(s) to your cart.`
> 
> Shopping cart & Product page (when this combination already NOT exists in cart): `There are only X product(s) left in stock.`

This brings another problem with plural versions of `product(s)` for different languages (if only it was as simple as english 😓). We should use sentences that avoid this problem. That's why I came up with:
`The quantity of this product available for purchase is X. You can add up to Y more to your cart.`
Simple numbers, no plural versions used.

**Comment by Julievrz** [comment]
We can use 2 strings (one singular and one plural). This way, it will be easier to translate. 

So, based on your feedback, we can go with:

**1. The user added more products than were available:**

- "There is only 1 product left in stock. Please adjust the quantity to proceed with your order."
- "There are only X products left. Please adjust the quantity to proceed with your order."

**2. The user added more products than were available AND a limit is set by the merchant:**

- "There is only 1 product left in stock. Please adjust the quantity to proceed with your order." (In this case, the limit has to be 1, right? So we can use the same string.)
- "You can add up to X products, within the limit of available stock (Y)."

So this would be 3 strings in total.

**Comment by SharakPL** [comment]
> We can use 2 strings (one singular and one plural).

My point is [we shouldn't](https://github.com/PrestaShop/hummingbird/issues/337#issuecomment-1179564181) 😉

https://github.com/PrestaShop/PrestaShop/issues/15870

**Comment by Julievrz** [comment]
> > We can use 2 strings (one singular and one plural).
> 
> My point is [we shouldn't](https://github.com/PrestaShop/hummingbird/issues/337#issuecomment-1179564181) 😉
> 
> #15870

@jolelievre I think you had found a solution for managing plural, using separators?

**Review by jolelievre — COMMENTED** [review]
Hi @Julievrz and @mparvazi 

the good news is that since 4.4 there is only one endpoint in the translator `trans` which is able to handle plural wordings so it might actually fix the problem automatically from this issue https://github.com/PrestaShop/PrestaShop/issues/15870 since we don't need to explicitly use `transChoice` anymore.

The bad news is that we detected a bug related to this that is being fixed in the PR https://github.com/PrestaShop/PrestaShop/pull/29040 but it's still a work in progress for now so it might block this PR until it's fixed.

But in any case, I think this is a good occasion to use plural wordings indeed.

**Comment by Julievrz** [comment]
> Hi @Julievrz and @mparvazi
> 
> the good news is that since 4.4 there is only one endpoint in the translator `trans` which is able to handle plural wordings so it might actually fix the problem automatically from this issue #15870 since we don't need to explicitly use `transChoice` anymore.
> 
> The bad news is that we detected a bug related to this that is being fixed in the PR #29040 but it's still a work in progress for now so it might block this PR until it's fixed.
> 
> But in any case, I think this is a good occasion to use plural wordings indeed.

Thanks for your answer. I agree that it would be the perfect occasion to try this new way of handling plural.

File: controllers/front/CartController.php
Comment (Julievrz) [inline controllers/front/CartController.php:439]:
```suggestion
                'The available order quantity for this product is %quantity%.',
```

**Review by marsaldev — APPROVED** [review]

**Comment by nicosomb** [comment]
As discussed with @NeOMakinG, this PR is ok for me. 

Steps: 
* merge this one 
* create a new issue to improve the displayed message (I looked a little bit to do it from this PR but the translator didn't accept the plural and I didn't find why).

**Comment by NeOMakinG** [comment]
I agree we should unblock this even without the translation problem

**Review by nicosomb — APPROVED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @mparvazi ,

Link to green automated tests ✅ 
https://github.com/florine2623/testing_pr/actions/runs/3931445255

Classic theme ✅ 

Hummingbird theme :x:
I don't have the same behavior as you when testing the Product Page in FO.
My minimum quantity is 3.
I can't add 4 products to my cart, but there's no toast or warning:

https://user-images.githubusercontent.com/16019289/212731771-f6d0ac24-7330-4b11-a958-f17275d0802d.mov

I should have something like this:
Screenshot 2023-01-16 at 18 01 55

Could you check ?

**Comment by Hlavtox** [comment]
@NeOMakinG Can you check this?

**Comment by NeOMakinG** [comment]
@florine2623 nothing related, this PR is working as expected, I did create a PR on hummingbird to handle these errors: https://github.com/PrestaShop/hummingbird/pull/435

I guess it's QA ✅  then as it was the only point, taking the liberty to merge it!

**Comment by SharakPL** [comment]
This file still needs some work because a lot isn't working properly.
For starters adding simple product to the cart and then changing this product in admin to product with combinations doesn't trigger this alert:
https://github.com/PrestaShop/PrestaShop/blob/329b763235d174458de3a81d633e5f4af723da67/controllers/front/CartController.php#L622-L628

Another one is that only one alert can be visible at a time, so when there are multiple errors with the cart and the one visible gets fixed the cart looks like everything is ok (no alerts) when in fact there are still errors.

Customisable products also don't work as expected. When you have one in your cart then sure enough you get the alert when you hit max quantity (eg. 30). But then you can go to that product's page, setup another customisation (even the same one as before) and you're still allowed to add another 30.

I'll create a proper issue for this in the morning...

**Pull request merged**
